gk/fib.c

```diff
@@ -0,0 +1,1640 @@
+
+			destroy_neigh_hash_table(&fib[0]->u.neigh);
+			initialize_fib_entry(fib[0]);
+			return -1;
+		}
+	}
+
+	/* Set up the FIB entry for the IPv6 network prefix. */
+	if (ipv6_if_configured(iface)) {
+		fib_id = get_empty_fib_id(ETHER_TYPE_IPv6, gk_conf);
+		if (fib_id < 0)
+			return -1;
+
+		fib[1] = &ltbl->fib_tbl6[fib_id];
+
+		if (iface == &net_conf->front) {
+			fib[1]->action = GK_FWD_NEIGHBOR_FRONT_NET;
+			gk_conf->neigh6_fib_front = fib[1];
+		} else if (likely(iface == &net_conf->back)) {
+			fib[1]->action = GK_FWD_NEIGHBOR_BACK_NET;
+			gk_conf->neigh6_fib_back = fib[1];
+		} else
+			rte_panic("Unexpected condition at %s: invalid interface %s!\n",
+				__func__, iface->name);
+
+		ret = setup_neighbor_tbl(socket_id, (identifier * 2 + 1),
+			ETHER_TYPE_IPv6, gk_conf->max_num_ipv6_neighbors,
+			&fib[1]->u.neigh6);
+		if (ret < 0) {
+			if (iface == &net_conf->front)
+				gk_conf->neigh6_fib_front = NULL;
+			else
+				gk_conf->neigh6_fib_back = NULL;
+
+			initialize_fib_entry(fib[1]);
+			goto free_fib0;
+		}
+
+		ret = gk_lpm_add_ipv6_route(iface->ip6_addr.s6_addr,
+			iface->ip6_addr_plen, fib_id, ltbl);
+		if (ret < 0) {
+			if (iface == &net_conf->front)
+				gk_conf->neigh6_fib_front = NULL;
+			else
+				gk_conf->neigh6_fib_back = NULL;
+
+			destroy_neigh_hash_table(&fib[1]->u.neigh);
+			initialize_fib_entry(fib[1]);
+			goto free_fib0;
+		}
+	}
+
+	return 0;
+
+free_fib0:
+	if (fib[0] != NULL) {
+		if (iface == &net_conf->front)
+			gk_conf->neigh_fib_front = NULL;
+		else
+			gk_conf->neigh_fib_back = NULL;
+
+		rte_lpm_delete(ltbl->lpm, ntohl(iface->ip4_addr.s_addr),
+			iface->ip4_addr_plen);
+		destroy_neigh_hash_table(&fib[0]->u.neigh);
+		initialize_fib_entry(fib[0]);
+	}
+
+	return -1;
+}
+
+static int
+init_fib_tbl(struct gk_config *gk_conf)
+{
+	int ret;
+	unsigned int i;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	rte_spinlock_init(&ltbl->lock);
+
+	for (i = 0; i < gk_conf->gk_max_num_ipv4_fib_entries; i++) {
+		ltbl->fib_tbl[i].action = GK_FIB_MAX;
+		rte_atomic16_init(&ltbl->fib_tbl[i].num_updated_instances);
+	}
+
+	for (i = 0; i < gk_conf->gk_max_num_ipv6_fib_entries; i++) {
+		ltbl->fib_tbl6[i].action = GK_FIB_MAX;
+		rte_atomic16_init(&ltbl->fib_tbl6[i].num_updated_instances);
+	}
+
+	/* Set up the FIB entry for the front network prefixes. */
+	ret = setup_net_prefix_fib(0, &gk_conf->net->front, gk_conf);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to setup the FIB entry for the front network prefixes at %s\n",
+			__func__);
+		goto out;
+	}
+
+	/* Set up the FIB entry for the back network prefixes. */
+	if (gk_conf->net->back_iface_enabled) {
+		ret = setup_net_prefix_fib(1, &gk_conf->net->back, gk_conf);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: failed to setup the FIB entry for the back network prefixes at %s\n",
+				__func__);
+			goto free_front_fibs;
+		}
+	}
+
+	ret = 0;
+	goto out;
+
+free_front_fibs:
+	if (gk_conf->neigh_fib_front != NULL) {
```

Comment:
If we decide to drop `gk_conf->neigh_fib_front` and `gk_conf->neigh6_fib_front` as disccused in another comment, the solution to keep this code working is for `setup_net_prefix_fib()` to return those values using parameters passed by reference.